services/motion/builtin/builtin.go

```diff
@@ -231,25 +235,149 @@ func (ms *builtIn) MoveOnGlobe(
 	kinematicsOptions.GoalRadiusMM = math.Min(motionCfg.PlanDeviationM*1000, 3000)
 	kinematicsOptions.HeadingThresholdDegrees = 8
 
-	plan, kb, err := ms.planMoveOnGlobe(
-		ctx,
-		componentName,
-		destination,
-		movementSensorName,
-		obstacles,
-		kinematicsOptions,
-		extra,
-	)
-	if err != nil {
-		return false, fmt.Errorf("error making plan for MoveOnMap: %w", err)
-	}
+	movementSensor, ok := ms.movementSensors[movementSensorName]
+	if !ok {
+		return false, resource.DependencyNotFoundError(movementSensorName)
+	}
+
+	var planMu sync.Mutex
+	var currentPlan [][]referenceframe.Input = nil
+	// TODO: if we disentangle creation of the kinematic base from the plan call we make we don't need this and can do this once
+	var kinematicBase kinematicbase.KinematicBase = nil
+
+	var positionMu sync.Mutex
+	// TODO: this has got to become a geo Point eventually but testing gets hard
+	var currentPosition *geo.Point
+
+	var replanFlag atomic.Bool
+	replanFlag.Store(true)
+
+	var successFlag atomic.Bool
+	successFlag.Store(false)
+
+	errChan := make(chan error)
+
+	cancelCtx, cancelFn := context.WithCancel(context.Background())
+	ms.cancelFn = cancelFn
+
+	positionPollingPeriod := time.Duration(1000/motionCfg.PositionPollingFreqHz) * time.Millisecond
+	obstaclePollingPeriod := time.Duration(1000/motionCfg.ObstaclePollingFreqHz) * time.Millisecond
+
+	startPolling := func(ctx context.Context, period time.Duration, fn func(context.Context)) {
+		ms.backgroundWorkers.Add(1)
+		goutils.ManagedGo(func() {
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					startTime := time.Now()
+					fn(ctx)
+					timeElapsed := time.Since(startTime)
+					sleepTime := int(math.Max(0, float64(period-timeElapsed)))
+					if !goutils.SelectContextOrWait(ctx, time.Duration(sleepTime)) {
+						fmt.Println("cancelled")
+						return
+					}
+				}
+			}
+		}, ms.backgroundWorkers.Done)
+	}
+
+	ms.backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		for !successFlag.Load() && ctx.Err() == nil {
+			if replanFlag.Load() {
+				fmt.Println("replanning")
+
+				planMu.Lock()
+				currentPlan = nil
+				planMu.Unlock()
+
+				ms.cancelFn()
```

Comment:
This is a race condition. The plan execution loop below may try to subset the now-nil plan and panic before being cancelled.

This line should move upwards above the `planMu.Lock`, and anything below which reads `currentPlan` should do a context check immediately after its own `planMu.Lock`, prior to actually reading the variable.

Reply:
I pretty drastically changed the way this is working and I think the new way avoids this issue, but am interested to get a second set of eyes on it